### PR TITLE
FB3651: Fix audio devices switching when they shouldn't

### PR DESCRIPTION
**Pull request by zfox23**
Also, I cleaned up the script and added better logging (in case we have to debug an issue with device selection again (the likelihood of which is almost 100%)).

**Test Plan:**
Generally, just try to break audio device selection. See if you can achieve unexpected behavior with respect to audio input/output devices.

Some things to try:
- Start Interface both in HMD mode and not in HMD mode
- Unplug the audio input device that you're using while Interface is running
- Unplug the audio input device that you're _not_ using while Interface is running
- Unplug the audio output device that you're using while Interface is running
- Unplug the audio output device that you're _not_ using while Interface is running
- Go into HMD mode, then back out of HMD mode, ensuring that the audio device switches back to what it was when you were in Desktop mode
- Go into HMD mode while you already have HMD audio devices selected
- Move between domains

In all of the above cases, make sure that your Audio menu displays the checked device correctly (both at the top of the Interface window and in the Tablet).

NOTE that the "Mute Microphone" option toggles in some cases while switching audio devices. This is not a result of this script and should be considered intended.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 31ae326880b18ff9754989557c4b1990c7ff9914 available at: https://gist.github.com/9f9042b323adb03119df0243d8b864ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-31ae326.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-31ae326.dmg

**Comment by zfox23** [comment]
Working with @kencooke - Not quite ready for prime-time yet.

**Comment by kencooke** [comment]
When I unplug the active output, I see several issues:

1) Warnings in my log:
```
[03/24 13:27:08] [WARNING] [default] Unable to find MenuUserData for object  QAction(0x2b1ac4bc330 text="" menuRole=TextHeuristicRole visible=true)
[03/24 13:27:08] [WARNING] [default] ""
[03/24 13:27:08] [WARNING] [default] Attempted to remove menu action with no found QML object
```

2) When I plug back in, I am unable to use the device. Switching via menu doesn't work.

3) Menu then becomes corrupt:
image

**Comment by zzmp** [comment]
Your comments (both the `/**/` style and "line break" style) are against [the coding standard](https://wiki.highfidelity.com/wiki/Coding_Standards#4.3_Comments).

**Comment by hifi-gustavo** [comment]
Build console output for commit: 73a77f7f2563c37f8831c751e2d3466cd5011d4d available at: https://gist.github.com/9f9042b323adb03119df0243d8b864ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-73a77f7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-73a77f7.dmg

File: scripts/system/selectAudioDevice.js [hunk 1]
```diff
@@ -45,65 +45,73 @@ if (typeof String.prototype.trimEndsWith != 'function') {
...
 }
 
+/****************************************
+    VAR DEFINITIONS
+****************************************/
 const INPUT_DEVICE_SETTING = "audio_input_device";
 const OUTPUT_DEVICE_SETTING = "audio_output_device";
+var audioDevicesList = [];
+// Some HMDs (like Oculus CV1) have a built in audio device. If they
+// do, then this function will handle switching to that device automatically
```
Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: this isn't a function - this comment seems to be in the wrong place

Comment (zfox23) [inline scripts/system/selectAudioDevice.js]: Old comment using unclear language, so I just missed cleaning this one up. Thanks.

File: scripts/system/selectAudioDevice.js [hunk 2]
```diff
@@ -45,65 +45,73 @@ if (typeof String.prototype.trimEndsWith != 'function') {
     };
 }
 
+/****************************************
+    VAR DEFINITIONS
+****************************************/
 const INPUT_DEVICE_SETTING = "audio_input_device";
 const OUTPUT_DEVICE_SETTING = "audio_output_device";
+var audioDevicesList = [];
```
Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: You can put all of these booleans into a `Settings` object to better organize it.

File: scripts/system/selectAudioDevice.js [hunk 3]
```diff
@@ -45,65 +45,73 @@ if (typeof String.prototype.trimEndsWith != 'function') {
...
+/****************************************
+    BEGIN FUNCTION DEFINITIONS
+****************************************/
 function setupAudioMenus() {
-    removeAudioMenus();
-    Menu.addSeparator("Audio", "Input Audio Device");
+    Menu.menuItemEvent.disconnect(menuItemEvent);
 
-    var inputDeviceSetting = Settings.getValue(INPUT_DEVICE_SETTING);
+    /* Setup audio input devices */
```
Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: All comments should be `//` type.

File: scripts/system/selectAudioDevice.js [hunk 4]
```diff
@@ -45,65 +45,73 @@ if (typeof String.prototype.trimEndsWith != 'function') {
...
-    var selectedInputDevice = AudioDevice.getInputDevice();
-    if (inputDevices.indexOf(inputDeviceSetting) != -1 && selectedInputDevice != inputDeviceSetting) {
-        if (AudioDevice.setInputDevice(inputDeviceSetting)) {
-            selectedInputDevice = inputDeviceSetting;
-        }
-    }
-    print("audio input devices: " + inputDevices);
-    for(var i = 0; i < inputDevices.length; i++) {
-        var thisDeviceSelected = (inputDevices[i] == selectedInputDevice);
+    print("selectAudioDevice: Audio input devices: " + inputDevices);
```
Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: Does this need to be printed?

File: scripts/system/selectAudioDevice.js [hunk 5]
```diff
@@ -45,65 +45,73 @@ if (typeof String.prototype.trimEndsWith != 'function') {
...
+    for (var i = 0; i < inputDevices.length; i++) {
         var menuItem = "Use " + inputDevices[i] + " for Input";
         Menu.addMenuItem({
             menuName: "Audio",
             menuItemName: menuItem,
             isCheckable: true,
-            isChecked: thisDeviceSelected
+            isChecked: inputDevices[i] == AudioDevice.getInputDevice()
```
Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: Why are you checking this every time instead of just once outside of the loop? This seems suboptimal - add a comment justifying it, or move it back out.

File: scripts/system/selectAudioDevice.js [hunk 6]
```diff
@@ -45,65 +45,73 @@ if (typeof String.prototype.trimEndsWith != 'function') {
...
         });
         audioDevicesList.push(menuItem);
-        if (thisDeviceSelected) {
-            selectedInputMenu = menuItem;
-        }
     }
 
+    /* Setup audio output devices */
```
Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: `//`

Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: This duplicates the setup of input devices - if you want to go crazy you can have one function that does both, and just takes "input" or "output" as an argument.

File: scripts/system/selectAudioDevice.js [hunk 7]
```diff
@@ -112,130 +120,140 @@ function removeAudioMenus() {
...
-    print("now setting up Audio > Devices menu");
-    setupAudioMenus();
-}
-
-// Have a small delay before the menu's get setup and the audio devices can switch to the last selected ones
-Script.setTimeout(function () {
-    print("connecting deviceChanged");
-    AudioDevice.deviceChanged.connect(onDevicechanged);
-    print("setting up Audio > Devices menu for first time");
+    print("selectAudioDevice: System audio device changed. Removing and replacing Audio Menus...");
```
Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: Do these debug statements need to be in here? Can they be toggled off in release?

File: scripts/system/selectAudioDevice.js [hunk 8]
```diff
@@ -112,130 +120,140 @@ function removeAudioMenus() {
...
+    removeAudioMenus();
```
Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: If `removeAudioMenus` is idempotent, can it be bundled with `setupAudioMenus` to reduce mental strain (i.e. developers won't need to remember to call both in some situations, because they will always be called together).

File: scripts/system/selectAudioDevice.js [hunk 9]
```diff
@@ -45,65 +45,73 @@ if (typeof String.prototype.trimEndsWith != 'function') {
...
+    }
+
+    Menu.menuItemEvent.connect(menuItemEvent);
+}
+
+function checkDeviceMismatch() {
+    var inputDeviceSetting = Settings.getValue(INPUT_DEVICE_SETTING);
+    var interfaceInputDevice = AudioDevice.getInputDevice();
+    if (interfaceInputDevice != inputDeviceSetting) {
+        print("selectAudioDevice: Input Setting & Device mismatch! Input SETTING:", inputDeviceSetting, "Input DEVICE IN USE:", AudioDevice.getInputDevice());
```
Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: I *love* that you are namespacing the comment so that it is obvious which script is printing it, but can you do it everywhere? Have a helper function `debug() { print.apply(["selectAudioDevice.js"].concat(arguments.slice())); }` or similar.

File: scripts/system/selectAudioDevice.js [hunk 10]
```diff
@@ -112,130 +120,140 @@ function removeAudioMenus() {
...
+            print("selectAudioDevice: User selected a new Audio Output Device: " + selectedDevice);
+            switchAudioDevice(false, selectedDevice);
+        } else {
+            print("selectAudioDevice: Invalid Audio menuItem! Doesn't end with 'for Input' or 'for Output'")
         }
     }
 }
 
-Menu.menuItemEvent.connect(menuItemEvent);
+function switchAudioDevice(isInput, device) {
```
Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: When invoking this, it is not at all obvious that a boolean corresponds to `isInput`. It makes it harder to read. Instead, can you have two separate functions that call into this (or wrap this with closure), `switchAudioInput` and `switchAudioOutput`?

Comment (zzmp) [inline scripts/system/selectAudioDevice.js]: In the same vein, ***much*** of this is repetition of input & output. Lots of the helper functions and logic can be rewritten to make use of this to improve clarity.

**Comment by kencooke** [comment]
I retested by unplugging and replugging the active audio device.

1) I still get the warnings in my log:
```
[03/24 18:48:03] [DEBUG] [hifi.scriptengine.script] script:print()<< selectAudioDevice: System audio device changed. Removing and replacing Audio Menus...
[03/24 18:48:03] [WARNING] [default] Unable to find MenuUserData for object  QAction(0x296a1893550 text="" menuRole=TextHeuristicRole visible=true)
[03/24 18:48:03] [WARNING] [default] ""
[03/24 18:48:03] [WARNING] [default] Attempted to remove menu action with no found QML object
[03/24 18:48:03] [WARNING] [default] Unable to find MenuUserData for object  QAction(0x296a15043b0 text="" menuRole=TextHeuristicRole visible=true)
[03/24 18:48:03] [WARNING] [default] ""
[03/24 18:48:03] [WARNING] [default] Attempted to remove menu action with no found QML object
[03/24 18:48:03] [DEBUG] [hifi.scriptengine.script] script:print()<< selectAudioDevice: Audio input devices: Digital Audio (S/PDIF) (High Definition Audio Device),Microphone (HD Webcam C615)
[03/24 18:48:03] [DEBUG] [hifi.scriptengine.script] script:print()<< selectAudioDevice: Audio output devices: Digital Audio (S/PDIF) (High Definition Audio Device),Digital Audio (S/PDIF) (High Definition Audio Device)
```
2) When I replug, the device shows up and is selected in the menu. But it does not work until I deselect and reselect it.

3) I still get garbage entries at the top of my menu:
image

**Comment by zfox23** [comment]
@kencooke:

Shoot. 😞 I'm pretty sure at least part of our problem here indeed has to do with you having two SPDIF outputs named the same thing. How has this worked before...? I'm positive that the original code wasn't thinking about the case where the strings weren't unique, but I guess my code exposes this problem...

In my thorough tests, I didn't get the logspam or the garbage menu entries, so I'm going to have to think some more about this...

When "thinking some more about this," I'll also take @zzmp's comments into considerations - thanks for your thoughtful review.

**Comment by ZappoMan** [comment]
Part of the problem is our JS menu API works on the string names used in the menus. 

One possible fix would be to change the menu API to return an ID that would then allow us to manage those menu calls with IDs as well as strings. 

Additionally we'd be able get audio devices by ID.

File: scripts/system/selectAudioDevice.js [hunk 11]
```diff
@@ -112,130 +120,140 @@ function removeAudioMenus() {
...
+    wasHmdActive = HMD.active;
 }
+/****************************************
+    END FUNCTION DEFINITIONS
+****************************************/
 
-Script.update.connect(checkHMDAudio);
+/****************************************
+    BEGIN SCRIPT BODY
+****************************************/
```
Comment (thoys) [inline scripts/system/selectAudioDevice.js]: Don't you wan't and END comment for this area below for consistency ? :)

**Comment by zfox23** [comment]
@kencooke: I investigated your three problems.

1. These warnings exist due to the way that VrMenu.cpp is written. This is not a problem specific to selectAudioDevice.js. I'm not going to spend time trying to figure out how to fix VrMenu.cpp.

I'm pretty sure these warnings would come up if a script tried to remove a checked menu item from any menu, not just the audio menu.

2. This is the way unplugging/replugging devices has always worked. I can work to fix it, but it feels like a lower priority to me.

3. This one feels like the highest priority to me. I don't see it happen on my machine, but I'm guessing this has to do with the duplicated device name. I will take a look at it.

**Comment by kencooke** [comment]
@zfox23 I agree, only #3 seems like a serious problem. Even that could be deferred to a later PR, if it works reliably for a typical setup.

File: scripts/system/selectAudioDevice.js (hunk 5 above)
Comment (zfox23) [inline scripts/system/selectAudioDevice.js]: I'm confused - this was never done outside the loop before. I just removed an unnecessary `var`

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0ae4ce7e985c86f7a600034c29ef176b84493c3a available at: https://gist.github.com/9f9042b323adb03119df0243d8b864ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-0ae4ce7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-0ae4ce7.dmg

**Comment by zfox23** [comment]
So..............I have been struggling with this for way too long. I feel like I've solved the problem, and then another one rears its ugly head.

The latest problem is that SOMETHING is making "phantom" menu selections in the Audio Devices menu when it shouldn't be. I've introduced a guard to attempt to prevent this from happening, and it's *mostly* successful, but it's still a huge problem. The phantom is selecting menu options and actually changing the active audio I/O devices after certain events (like switching from HMD to desktop - but only SOMETIMES).

I'm wondering if this is happening because of the way the Tablet Settings menu was programmed. I know for a fact that some bindings in there are broken - thus the `WARNINGS` that @kencooke is seeing in his logs above.

In other words, **I've fixed many problems with this script, but not the one that caused me to look at it in the first place**. Thus, I'd like to pass the torch off to someone else, please - I need more eyes on this problem. There are lots of new, detailed logs in the script to help the next person with debugging.

@kencooke: I fixed Problem #2 from above in my current iteration of the script. You can now unplug a selected device, then plug it back in, and Interface will reinitialize it.

**Comment by zzmp** [comment]
@kencooke can you capture your log (you can filter it by "selectAudioDevice.js") while your menu gets garbaged up? I'm unable to repro that on my own, and hoping that a log would help me get somewhere with it.

I've also added a few commits that will *hopefully* clean up logging a bit.

**Comment by zzmp** [comment]
And I am trying to repro the original switching audio issue on my own - will post results when I have them.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 06bf5807ae3a0017832db495616d85a518da35e0 available at: https://gist.github.com/9f9042b323adb03119df0243d8b864ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-06bf580.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-06bf580.dmg

**Comment by kencooke** [comment]
@zzmp this is when I unplug the active device:
```
[03/28 13:12:57] [DEBUG] [hifi.scriptengine.script] script:print()<< selectAudioDevice.js: "System audio devices changed. Removing and replacing Audio Menus..."
[03/28 13:12:57] [WARNING] [default] Unable to find MenuUserData for object  QAction(0x19d10fe9d10 text="" menuRole=TextHeuristicRole visible=true)
[03/28 13:12:57] [WARNING] [default] ""
[03/28 13:12:57] [WARNING] [default] Attempted to remove menu action with no found QML object
[03/28 13:12:57] [WARNING] [default] Unable to find MenuUserData for object  QAction(0x19d10fea070 text="" menuRole=TextHeuristicRole visible=true)
[03/28 13:12:57] [WARNING] [default] ""
[03/28 13:12:57] [WARNING] [default] Attempted to remove menu action with no found QML object
[03/28 13:12:57] [DEBUG] [hifi.audioclient] [ "Speakers (RME Babyface)" ] [ "" ]
[03/28 13:12:57] [DEBUG] [hifi.audioclient] [ "Analog (1+2) (RME Babyface)" ] [ "" ]
[03/28 13:12:57] [DEBUG] [hifi.scriptengine.script] script:print()<< selectAudioDevice.js: "Input Setting & Device mismatch! Input SETTING: Analog (1+2) (RME Babyface)Input DEVICE IN USE: "
```

and this is when I plug it back in:
```
[03/28 13:14:54] [DEBUG] [hifi.scriptengine.script] script:print()<< selectAudioDevice.js: "System audio devices changed. Removing and replacing Audio Menus..."
[03/28 13:14:54] [WARNING] [default] Unable to find MenuUserData for object  QAction(0x19d276dea60 text="" menuRole=TextHeuristicRole visible=true)
[03/28 13:14:54] [WARNING] [default] ""
[03/28 13:14:54] [WARNING] [default] Attempted to remove menu action with no found QML object
[03/28 13:14:54] [WARNING] [default] Unable to find MenuUserData for object  QAction(0x19d2c592af0 text="" menuRole=TextHeuristicRole visible=true)
[03/28 13:14:54] [WARNING] [default] ""
[03/28 13:14:54] [WARNING] [default] Attempted to remove menu action with no found QML object
[03/28 13:14:54] [DEBUG] [hifi.audioclient] [ "Speakers (RME Babyface)" ] [ "Speakers (RME Babyface)" ]
[03/28 13:14:54] [DEBUG] [hifi.audioclient] The audio output device  "Speakers (RME Babyface)" is available.
[03/28 13:14:54] [DEBUG] [hifi.audioclient] The desired format for audio I/O is QAudioFormat(24000Hz, 16bit, channelCount=2, sampleType=SignedInt, byteOrder=LittleEndian, codec="audio/pcm")
[03/28 13:14:54] [DEBUG] [hifi.audioclient] The format to be used for audio output is QAudioFormat(48000Hz, 16bit, channelCount=2, sampleType=SignedInt, byteOrder=LittleEndian, codec="audio/pcm")
[03/28 13:14:54] [DEBUG] [hifi.audioclient] Attemping to create a resampler for network format to output format.
[03/28 13:14:54] [DEBUG] [hifi.audioclient] Output Buffer capacity in frames:  2.2 requested bytes: 1920 actual bytes: 4224 os default: 4096 period size: 1920
[03/28 13:14:54] [DEBUG] [hifi.networking] unreasonable sequence number: 16650 previous: 4897
[03/28 13:14:54] [DEBUG] [hifi.audioclient] [ "" ] [ "" ]
[03/28 13:14:54] [DEBUG] [hifi.scriptengine.script] script:print()<< selectAudioDevice.js: "Input Setting & Device mismatch! Input SETTING: Analog (1+2) (RME Babyface)Input DEVICE IN USE: "
[03/28 13:14:54] [DEBUG] [hifi.networking] re-synced with sequence number sender
[03/28 13:14:54] [DEBUG] [hifi.scriptengine.script] script:print()<< selectAudioDevice.js: "Switching audio INPUT device from  to Analog (1+2) (RME Babyface)"
[03/28 13:14:54] [DEBUG] [hifi.audioclient] [ "Analog (1+2) (RME Babyface)" ] [ "Analog (1+2) (RME Babyface)" ]
[03/28 13:14:54] [DEBUG] [hifi.audioclient] The audio input device  "Analog (1+2) (RME Babyface)" is available.
[03/28 13:14:54] [DEBUG] [hifi.audioclient] The desired format for audio I/O is QAudioFormat(24000Hz, 16bit, channelCount=1, sampleType=SignedInt, byteOrder=LittleEndian, codec="audio/pcm")
[03/28 13:14:54] [DEBUG] [hifi.audioclient] The format to be used for audio input is QAudioFormat(48000Hz, 16bit, channelCount=2, sampleType=SignedInt, byteOrder=LittleEndian, codec="audio/pcm")
[03/28 13:14:54] [DEBUG] [hifi.audioclient] Attemping to create a resampler for input format to network format.
[03/28 13:14:54] [DEBUG] [hifi.scriptengine.script] script:print()<< selectAudioDevice.js: "Selected input device is the same as the current input device!"
[03/28 13:14:54] [DEBUG] [hifi.audioclient] [ "Analog (1+2) (RME Babyface)" ] [ "Analog (1+2) (RME Babyface)" ]
[03/28 13:14:54] [DEBUG] [hifi.audioclient] The audio input device  "Analog (1+2) (RME Babyface)" is available.
[03/28 13:14:54] [DEBUG] [hifi.audioclient] The desired format for audio I/O is QAudioFormat(24000Hz, 16bit, channelCount=1, sampleType=SignedInt, byteOrder=LittleEndian, codec="audio/pcm")
[03/28 13:14:54] [DEBUG] [hifi.audioclient] The format to be used for audio input is QAudioFormat(48000Hz, 16bit, channelCount=2, sampleType=SignedInt, byteOrder=LittleEndian, codec="audio/pcm")
[03/28 13:14:54] [DEBUG] [hifi.audioclient] Attemping to create a resampler for input format to network format.
[03/28 13:14:55] [DEBUG] [hifi.audioclient] Starve detected, 1 new unfulfilled reads
```

**Comment by zfox23** [comment]
@kencooke: these logs make sense - does the audio device then work afterwards? it should.

Some of the warnings you're seeing are warnings neither @zzmp nor myself have attempted to fix - they've been around for a long time, and shouldn't hinder audio device selection or operation.

**Comment by zzmp** [comment]
@kencooke @zfox23 could you test this again on your setups?

I've added a 200ms delay instead of disconnecting and reconnecting the Qt signals for Men.menuItemEvent, because of [what appears to be a threading bug](https://highfidelity.fogbugz.com/f/cases/3745/JavaScript-events-triggered-while-disconnected-may-still-be-delivered). Although it's not the cleanest, I believe it will work in all cases.

**Comment by hifi-gustavo** [comment]
Build console output for commit: f514c200ef7b8cd9e81fb2d30ad4e1506675ca82 available at: https://gist.github.com/9f9042b323adb03119df0243d8b864ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-f514c20.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-f514c20.dmg

**Comment by zfox23** [comment]
Testing now...

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4a9fbf2bbd5d7820b39e1efa637c9d344de3a14d available at: https://gist.github.com/9f9042b323adb03119df0243d8b864ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-4a9fbf2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9996-4a9fbf2.dmg

**Comment by zfox23** [comment]
Yep, at first pass, I'm no longer running into my previous issues. @kencooke has had better luck breaking this script, so hopefully he can take a whack at it. Awesome stuff, @zzmp. Thanks for taking over the reins on this one.

**Comment by kencooke** [comment]
The important functionality works for me.

I do still get the warnings in my log -- one for each corrupt entry in my menu (see image below). I'm sure these are caused by my system having two devices with the identical name string. We can defer this to a later PR, since it requires changes beyond just selectAudioDevices.js. As zappoman said, we need the IDs or to append "(1)" and "(2)" to the strings or something.

selectaudiodevice

**Comment by Nex-Pro** [comment]
#9996

Testing platform: Intel workstation.
OS: Win10 pro 64
CPU: i7 4.0 ghz
RAM: 16gb
GPU: Nvidia GTX1070 8gb VRam
HMD: Htc Vive.

Audio systems tested:  USB headset + Vive audio.


Test Plan:
Generally, just try to break audio device selection.
See if you can achieve unexpected behavior with respect to audio input/output devices.

Some things to try:

Start Interface both in HMD mode and not in HMD mode:
    **No issues seen**

Unplug the audio input device that you're using while Interface is running:
    **Unplugged my usual USB audio headset.**
        Interface then showed no selected audio device but did not crash.

Unplug the audio input device that you're not using while Interface is running:
    **Had vive audio seleced then unplugged my usual USB audio headset.**
        The unplugged audio device was no longer shown in the audio menu.
        Next I plugged the usb headset back in.
        Interface audio menu updated and displayed my usual USB audio headset as an option after plugging it back in.


Go into HMD mode, then back out of HMD mode, 
ensuring that the audio device switches back to what it was when you were in Desktop mode: 
    **Does not change between modes. Remains on whatever option was last selected.**

Go into HMD mode while you already have HMD audio devices selected:**DONE**

Move between domains: 
    **No issues seen**

In all of the above cases, 
make sure that your Audio menu displays the checked device correctly : 
    **SUCCESS**
(both at the top of the Interface window and in the Tablet): 
    **SUCCESS**


NOTE that the "Mute Microphone" option toggles in some cases while switching audio devices. 
This is not a result of this script and should be considered intended: **NOTED**


### Observation:

    It is possible to select the on-board audio and vive or usb at the same time. Not Vive and USB. See image.

    Audio menu can be corrupted as noted by Ken.

audiomenufreshopened

audiomenu

**Comment by zzmp** [comment]
You two both seem to have to "Digital Audio (S/PDIF) (High Definition Audio Device)".
Is there a way to have my system think that I have two of those, so that we can also repro the bad menu issues?

**Comment by zzmp** [comment]
CR - if you approve this, go ahead and click merge :)
